From 38b5ac58be2a0b46adc97f36c54da80e1a54d0e3 Mon Sep 17 00:00:00 2001
From: b02247 <b02247@freescale.com>
Date: Wed, 15 Aug 2012 17:26:07 +0800
Subject: [PATCH 329/486] ENGR00219871 audio hal: implement the api:
 in_get_input_frames_lost

implement the api:in_get_input_frames_lost

Signed-off-by: b02247 <b02247@freescale.com>
---
 alsa/audio_hardware.h | 1 +
 alsa/tinyalsa_hal.c   | 9 ++++++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/alsa/audio_hardware.h b/alsa/audio_hardware.h
index 5a0a82d..f339036 100644
--- a/alsa/audio_hardware.h
+++ b/alsa/audio_hardware.h
@@ -137,6 +137,7 @@ struct imx_stream_in {
     int read_status;
     size_t mute_500ms;
     struct imx_audio_device *dev;
+    int last_time_of_xrun;
 };
 
 #define SUPPORTED_DEVICE_IN_MODULE               \
diff --git a/alsa/tinyalsa_hal.c b/alsa/tinyalsa_hal.c
index 8b0bddd..0bf485e 100644
--- a/alsa/tinyalsa_hal.c
+++ b/alsa/tinyalsa_hal.c
@@ -1037,6 +1037,7 @@ static int do_input_standby(struct imx_stream_in *in)
         LOGW("do_in_standby..");
         pcm_close(in->pcm);
         in->pcm = NULL;
+        in->last_time_of_xrun = 0;
 
         adev->active_input = 0;
         if (adev->mode != AUDIO_MODE_IN_CALL) {
@@ -1497,7 +1498,13 @@ exit:
 
 static uint32_t in_get_input_frames_lost(struct audio_stream_in *stream)
 {
-    return 0;
+    int times, diff;
+    struct imx_stream_in *in = (struct imx_stream_in *)stream;
+    times = pcm_get_time_of_xrun(in->pcm);
+    diff = times - in->last_time_of_xrun;
+    LOGW_IF((diff != 0), "in_get_input_frames_lost %d ms total %d ms\n",diff, times);
+    in->last_time_of_xrun = times;
+    return diff * in->requested_rate / 1000;
 }
 
 static int in_add_audio_effect(const struct audio_stream *stream,
-- 
1.8.0

